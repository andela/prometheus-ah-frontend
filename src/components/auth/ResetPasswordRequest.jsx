import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { startResetPassword } from '../../actions/authentication/resetPassword';
import resetPasswordValidations from '../../validations/resetPasswordValidation';
import routes from '../../constants/routes';
import TextField from '../common/TextField';


/**
 * sign in an existing user
 *
 * @class ResetPasswordRequest
 * @param {event} event

 * @extends {Component}
 */
class ResetPasswordRequest extends Component {
  state = {
    email: '',
    emailValid: false,
    error: {},
  };

  onChange = (event) => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };

  onSubmit = (event) => {
    event.preventDefault();

    const { error, hasErrors } = resetPasswordValidations.emailInputValidation(
      this.state
    );
    this.setState(() => ({
      error
    }));
    if (!hasErrors) {
      const { email } = this.state;
      const { resetPasswordRequest } = this.props;
      resetPasswordRequest(email)
        .then((payload) => {
          if (payload) {
            this.setState(() => ({
              emailValid: true
            }));
          }
        })
    }
  };

  /**
   * renders component to DOM
   *
   * @memberof PasswordReset
   *
   * @returns {JSX} JSX representation of component
   */
  render() {
    const { error, email, emailValid } = this.state;
    if (emailValid) {
      return <Redirect to={routes.LANDING} />;
    }
    return (
      <div className="form-gap">
        <div className="container">
          <div className="form-row text-center justify-content-center">
            <div className="col-md-6 col-md-offset-4 border rounded">
              <div className="panel panel-default">
                <div className="panel-body">
                  <div className="text-center">
                    <h3>
                      <i className="fa fa-lock fa-4x" />
                    </h3>
                    <h2 className="text-center">Forgot Password?</h2>
                    <p>Enter Your email address to get a password reset link</p>
                    <div className="panel-body">
                      <form
                        id="register-form"
                        autoComplete="off"
                        className="form"
                        method="post"
                        onSubmit={this.onSubmit}
                      >
                        <div className="form-group">
                          <TextField
                            error={error.email}
                            onChange={this.onChange}
                            value={email}
                            placeholder="enter your email"
                            type="email"
                            className="form-control form-input"
                            field="email"
                          />
                        </div>
                        <div className="form-group">
                          <input
                            name="recover-submit"
                            className="btn btn-lg btn-block login-btn mt-4"
                            value="Send email verification link"
                            type="submit"
                          />
                        </div>
                      </form>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

ResetPasswordRequest.defaultProps = {
  resetPasswordRequest: () => {}
};

ResetPasswordRequest.propTypes = {
  resetPasswordRequest: PropTypes.func
};

const mapDispatchToProps = dispatch => ({
  resetPasswordRequest: bindActionCreators(startResetPassword, dispatch),
});

export default connect(
  undefined,
  mapDispatchToProps
)(ResetPasswordRequest);
