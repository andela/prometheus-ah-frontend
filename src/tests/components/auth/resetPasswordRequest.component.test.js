import React from 'react';
import { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import ResetPasswordRequest from '../../../components/auth/ResetPasswordRequest';
import mockData from '../../__mocks__/mockData';

let resetPasswordRequest;
let wrapper;

const createMockStore = configureMockStore([thunk]);

describe('ResetPasswordRequest', () => {
  beforeEach(() => {
    resetPasswordRequest = jest.fn(() => Promise.resolve());
    wrapper = shallow(
      <ResetPasswordRequest.WrappedComponent
        resetPasswordRequest={resetPasswordRequest}
      />
    );
  });

  test('should correctly render ResetPasswordRequestPage ', () => {
    expect(wrapper).toMatchSnapshot();
  });
  test('should correctly render ResetPasswordRequestPage ', () => {
    wrapper.setState({ emailValid: true });
    expect(wrapper).toMatchSnapshot();
  });

  test('should set email on input change', () => {
    const { signUpDetails } = mockData;
    const name = 'email';
    const value = signUpDetails.email;
    wrapper
      .find('TextField')
      .simulate('change', {
        target: { name, value }
      });
    expect(wrapper.state(name)).toBe(value);
  });

  test('should return error for invalid form submission', () => {
    wrapper.find('form').simulate('submit', {
      preventDefault: () => {}
    });
    expect(wrapper.state('error')).not.toBe({});
    expect(wrapper).toMatchSnapshot();
  });

  test('should return error for invalid form submission', () => {
    resetPasswordRequest = jest.fn(() => Promise.reject());
    const wrapper3 = shallow(
      <ResetPasswordRequest.WrappedComponent resetPasswordRequest={resetPasswordRequest} />
    );
    wrapper3.find('form').simulate('submit', {
      preventDefault: () => {}
    });
    expect(wrapper3.state('error')).not.toBe({});
    expect(wrapper3).toMatchSnapshot();
  });

  test('should call resetPasswordRequest props for valid form submission', () => {
    const { signUpDetails } = mockData;
    const data = { email: signUpDetails.email };
    wrapper.setState(data);
    wrapper.find('form').simulate('submit', {
      preventDefault: () => {}
    });
    expect(resetPasswordRequest).toHaveBeenCalledWith(signUpDetails.email);
    expect(wrapper).toMatchSnapshot();
  });

  test('should call the dispatch mapped to the props', () => {
    const props = {
      resetPasswordRequest: jest.fn(() => Promise.resolve())
    };
    const store = createMockStore({});
    const wrapper2 = shallow(
      <ResetPasswordRequest store={store} {...props} />
    );
    expect(wrapper2).toMatchSnapshot();
  });
});
