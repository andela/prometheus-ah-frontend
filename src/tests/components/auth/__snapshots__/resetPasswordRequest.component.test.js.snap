// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetPasswordRequest should call resetPasswordRequest props for valid form submission 1`] = `
<div
  className="form-gap"
>
  <div
    className="container"
  >
    <div
      className="form-row text-center justify-content-center"
    >
      <div
        className="col-md-6 col-md-offset-4 border rounded"
      >
        <div
          className="panel panel-default"
        >
          <div
            className="panel-body"
          >
            <div
              className="text-center"
            >
              <h3>
                <i
                  className="fa fa-lock fa-4x"
                />
              </h3>
              <h2
                className="text-center"
              >
                Forgot Password?
              </h2>
              <p>
                Enter Your email address to get a password reset link
              </p>
              <div
                className="panel-body"
              >
                <form
                  autoComplete="off"
                  className="form"
                  id="register-form"
                  method="post"
                  onSubmit={[Function]}
                >
                  <div
                    className="form-group"
                  >
                    <TextField
                      className="form-control form-input"
                      field="email"
                      onChange={[Function]}
                      placeholder="enter your email"
                      type="email"
                      value="ugochukwu.ezeh@yahoo.com"
                    />
                  </div>
                  <div
                    className="form-group"
                  >
                    <input
                      className="btn btn-lg btn-block login-btn mt-4"
                      name="recover-submit"
                      type="submit"
                      value="Send email verification link"
                    />
                  </div>
                </form>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ResetPasswordRequest should call the dispatch mapped to the props 1`] = `
<ResetPasswordRequest
  resetPasswordRequest={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
/>
`;

exports[`ResetPasswordRequest should correctly render ResetPasswordRequestPage  1`] = `
<div
  className="form-gap"
>
  <div
    className="container"
  >
    <div
      className="form-row text-center justify-content-center"
    >
      <div
        className="col-md-6 col-md-offset-4 border rounded"
      >
        <div
          className="panel panel-default"
        >
          <div
            className="panel-body"
          >
            <div
              className="text-center"
            >
              <h3>
                <i
                  className="fa fa-lock fa-4x"
                />
              </h3>
              <h2
                className="text-center"
              >
                Forgot Password?
              </h2>
              <p>
                Enter Your email address to get a password reset link
              </p>
              <div
                className="panel-body"
              >
                <form
                  autoComplete="off"
                  className="form"
                  id="register-form"
                  method="post"
                  onSubmit={[Function]}
                >
                  <div
                    className="form-group"
                  >
                    <TextField
                      className="form-control form-input"
                      field="email"
                      onChange={[Function]}
                      placeholder="enter your email"
                      type="email"
                      value=""
                    />
                  </div>
                  <div
                    className="form-group"
                  >
                    <input
                      className="btn btn-lg btn-block login-btn mt-4"
                      name="recover-submit"
                      type="submit"
                      value="Send email verification link"
                    />
                  </div>
                </form>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ResetPasswordRequest should correctly render ResetPasswordRequestPage  2`] = `
<Redirect
  push={false}
  to="/"
/>
`;

exports[`ResetPasswordRequest should return error for invalid form submission 1`] = `
<div
  className="form-gap"
>
  <div
    className="container"
  >
    <div
      className="form-row text-center justify-content-center"
    >
      <div
        className="col-md-6 col-md-offset-4 border rounded"
      >
        <div
          className="panel panel-default"
        >
          <div
            className="panel-body"
          >
            <div
              className="text-center"
            >
              <h3>
                <i
                  className="fa fa-lock fa-4x"
                />
              </h3>
              <h2
                className="text-center"
              >
                Forgot Password?
              </h2>
              <p>
                Enter Your email address to get a password reset link
              </p>
              <div
                className="panel-body"
              >
                <form
                  autoComplete="off"
                  className="form"
                  id="register-form"
                  method="post"
                  onSubmit={[Function]}
                >
                  <div
                    className="form-group"
                  >
                    <TextField
                      className="form-control form-input"
                      error={
                        Array [
                          "This email field is required.",
                        ]
                      }
                      field="email"
                      onChange={[Function]}
                      placeholder="enter your email"
                      type="email"
                      value=""
                    />
                  </div>
                  <div
                    className="form-group"
                  >
                    <input
                      className="btn btn-lg btn-block login-btn mt-4"
                      name="recover-submit"
                      type="submit"
                      value="Send email verification link"
                    />
                  </div>
                </form>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ResetPasswordRequest should return error for invalid form submission 2`] = `
<div
  className="form-gap"
>
  <div
    className="container"
  >
    <div
      className="form-row text-center justify-content-center"
    >
      <div
        className="col-md-6 col-md-offset-4 border rounded"
      >
        <div
          className="panel panel-default"
        >
          <div
            className="panel-body"
          >
            <div
              className="text-center"
            >
              <h3>
                <i
                  className="fa fa-lock fa-4x"
                />
              </h3>
              <h2
                className="text-center"
              >
                Forgot Password?
              </h2>
              <p>
                Enter Your email address to get a password reset link
              </p>
              <div
                className="panel-body"
              >
                <form
                  autoComplete="off"
                  className="form"
                  id="register-form"
                  method="post"
                  onSubmit={[Function]}
                >
                  <div
                    className="form-group"
                  >
                    <TextField
                      className="form-control form-input"
                      error={
                        Array [
                          "This email field is required.",
                        ]
                      }
                      field="email"
                      onChange={[Function]}
                      placeholder="enter your email"
                      type="email"
                      value=""
                    />
                  </div>
                  <div
                    className="form-group"
                  >
                    <input
                      className="btn btn-lg btn-block login-btn mt-4"
                      name="recover-submit"
                      type="submit"
                      value="Send email verification link"
                    />
                  </div>
                </form>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
